### PR TITLE
u-turn env addition

### PR DESCRIPTION
**Pull request by ashishrana160796**
Hi @eleurent,

Sharing with you my implementation of a U-Turn driving scenario. Hope this environment contribution is considered under scope of this project. Also, reason for hard-cording vehicle positions were to study trade-offs between overtaking and lane cruising behavior. Here, I am attaching below a sample video from one of my test runs for reference.

Thanks
Ashish Rana

https://user-images.githubusercontent.com/19948632/107861172-fa022c80-6e69-11eb-84ae-db8426475447.mp4

### CODE REVIEW

**Comment by eleurent** [comment]
Hi @ashishrana160796,
Thank you for this very nice contribution!

I have a few comments:
- the controlled vehicle temporarily leaves the road, which does not look great. I think this could be addressed in two ways:
  * increase the lateral/steering control gains
  * (better) add a penalty for offroad driving in the reward function, which should incentivize the agent to slow down before the u-turn
- the collision between those two vehicles is quite suspicious, they don't seem to be overlapping so there is probably something wrong with the rectangle intersection code... I'll investigate

File: highway_env/envs/u_turn_env.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+    def _make_road(self, length=128):
+        """
+        Making double lane road with counter-clockwise U-Turn.
+        :return: the road
+        """
+        net = RoadNetwork()
+
+        # Defining lower starting lanes before the U-Turn.
+        net.add_lane("c", "d", StraightLane([length, 0], [0, 0],
```
Comment (eleurent) [inline highway_env/envs/u_turn_env.py:93]: Aren't the lanes from/to labels mixed up ?
I think it should be:
* lower segment (before turn): A -> B
* turn: B -> C
* upper segment (after turn): C -> D

But it seems to me that you swapped upper and lower segments... Not sure why it still works though ^^

**Review by eleurent — CHANGES_REQUESTED** [review]

File: highway_env/envs/u_turn_env.py (hunk 1 above)
Comment (ashishrana160796) [reply highway_env/envs/u_turn_env.py:93]: Yes, you are right in your findings. I realized it now there is a mistake in my added comments. I'll update those comments as they are misleading.

**Comment by ashishrana160796** [comment]
Sure @eleurent,

Adding penalty for off-road driving in the reward function looks good to me. I'll add the suggested changes by EOD. Thanks for the review.

Also, actually I wanted to simulate vehicle collisions in the different environments as well. But, I encountered a relatively unpredictable behavior like the one shown in above. Which essentially has kept my collision experimentation on hold.

**Comment by eleurent** [comment]
> Adding penalty for off-road driving in the reward function looks good to me. I'll add the suggested changes by EOD. Thanks for the review.

To that end, you can use `self.vehicle.on_road`

> Also, actually I wanted to simulate vehicle collisions in the different environments as well. But, I encountered a relatively unpredictable behavior like the one shown in above. Which essentially has kept my collision experimentation on hold.

I will see if I can reproduce and fix the problem.

**Comment by ashishrana160796** [comment]
So, i have added the on_road constraint into the reward function based on your shared inputs. Thanks!!

**Comment by eleurent** [comment]
Thanks a lot!

**Comment by eleurent** [comment]
I did some minor modifications, and the environment is now quite nice!
Here are two episodes with the OPD planner


https://user-images.githubusercontent.com/1706935/108052263-28d7f880-704c-11eb-9b64-ab87b41f944f.mp4

https://user-images.githubusercontent.com/1706935/108052282-2ecdd980-704c-11eb-908a-647958b6c36c.mp4

**Comment by ashishrana160796** [comment]
Thanks, for all the refactoring effort!! Can't believe there were so many typos 🤦🏽 from my end.
& Yea, the output actually looks better now. Thanks again!!

**Pull request merged**